instance/models/openedx_appserver.py

```diff
@@ -0,0 +1,231 @@
+        immutable once this AppServer is saved.
+        """
+        template = loader.get_template(self.CONFIGURATION_VARS_TEMPLATE)
+        vars_str = template.render({
+            'instance': self,  # TODO: Rename this template var to 'appserver'
+            # This property is needed twice in the template.  To avoid evaluating it twice (and
+            # querying the Github API twice), we pass it as a context variable.
+            'github_admin_username_list': self.instance.github_admin_username_list,
+        })
+        for attr_name in self.CONFIGURATION_EXTRA_FIELDS:
+            additional_vars = getattr(self, attr_name)
+            vars_str = ansible.yaml_merge(vars_str, additional_vars)
+        self.logger.debug('Vars.yml:\n%s', vars_str)
+        return vars_str
+
+    @log_exception
+    @AppServer.status.only_for(AppServer.Status.New)
+    def provision(self):
+        """
+        Provision this AppServer.
+
+        Returns True on success or False on failure
+        """
+        self.logger.info('Starting provisioning')
+        # Start by requesting a new server/VM:
+        self._status_to_waiting_for_server()
+        assert self.server_id is None
+        server_prefix = ('edxapp-' + slugify(self.instance.sub_domain))[:20]
+        self.server.create(name_prefix=server_prefix)
+        self.server.start()
+
+        def accepts_ssh_commands():
+            """ Does server accept SSH commands? """
+            return self.server.status.accepts_ssh_commands
+
+        try:
+            self.logger.info('Waiting for server %s...', self.server)
+            self.server.sleep_until(lambda: self.server.status.vm_available)
+            self.logger.info('Waiting for server %s to finish booting...', self.server)
+            self.server.sleep_until(accepts_ssh_commands)
+        except SteadyStateException:
+            self._status_to_error()
+            return False
+
+        try:
+            # Provisioning (ansible)
+            self.logger.info('Provisioning server...')
+            self._status_to_configuring_server()
+            log, exit_code = self.run_ansible_playbooks()
+            if exit_code != 0:
+                self.logger.info('Provisioning failed')
+                self._status_to_configuration_failed()
+                self.provision_failed_email("AppServer deploy failed: ansible play exited with non-zero exit code", log)
+                return False
+
+            # Reboot
+            self.logger.info('Provisioning completed')
+            self.logger.info('Rebooting server %s...', self.server)
+            self.server.reboot()
+            self.server.sleep_until(accepts_ssh_commands)
+
+            # Declare instance up and running
+            self._status_to_running()
+
+            return True
+
+        except:
+            self.server.terminate()
+            self.provision_failed_email("AppServer deploy failed: unhandled exception")
```

Comment:
Did you remove the retries from https://github.com/open-craft/opencraft/pull/63/files#diff-848aab5153020ca3817d6db9d4d3a144L546 ? From a quick look I couldn't find this in the new code.